raft/raft.hh

```diff
@@ -351,6 +351,19 @@ struct append_request {
     // Log entries to store (empty vector for heartbeat; may send more
     // than one entry for efficiency).
     std::vector<log_entry_ptr> entries;
+
+    append_request copy() const {
```

Comment:
I think this should be the copy constructor.

Reply:
It cannot be the copy constructor since it makes a "deep" copy by copying the data the entries point too. Copy constructor would just copy shared pointers.

Reply:
The default copy constructor copies shared pointers, which is (arguably) incorrect, I'm suggesting to provide a custom copy constructor to make this deep copy.

Reply:
If we want to keep the default copy constructor for some reason, then:
- rename this function to `deep_copy()`,
- write a warning comment (here in the struct) that the copy constructor does a shallow copy and is unsafe e.g. for cross-shard situations.

Reply:
> The default copy constructor copies shared pointers, which is (arguably) incorrect, I'm suggesting to provide a custom copy constructor to make this deep copy.

Why is it incorrect? It is correct for regular use. Cross shard copy is more of an exception.

Reply:
Deep copy is correct in all situations, shallow copy is a "dangerous" optimization that doesn't always work (cross-shard). I'd prefer we go with correctness by default, and optimizations should be explicitly requested when the caller guarantees they are safe, so I'd even go with having a deep copy in the copy constructor and a separate function (if needed) `shallow_copy()` for the optimized case. But meh, it's subjective.

Reply:
When one uses shared pointer it explicitly says he wants shallow copy. If you do not want shallow copy just use the type directly. In C++ default copy constructor does deep copying. We use shared pointer because we want its behaviour in most cases. When we do not (as in this case) we add a function to do something different.

Reply:
I tried to entirely disable copy ctor/operator, but then aggregate initialisation required for RPC serialisation code stops working and I'm not particularly motivated about researching it. To me, all this matters is entirely subjective and shouldn't block the PR.